File: 1861-rotating-the-box/solution.py
class Solution:
    def rotateTheBox(self, boxGrid: list[list[str]]) -> list[list[str]]:
        m = len(boxGrid)
        n = len(boxGrid[0])

        # 1. Simulate gravity row by row
        for row in boxGrid:
            # empty_pos points to the rightmost available cell where a stone can land
            empty_pos = n - 1
            for j in range(n - 1, -1, -1):
                if row[j] == '*':
                    # Obstacle found, the next stone must land to the left of it
                    empty_pos = j - 1
                elif row[j] == '#':
                    # Stone found, move it to the rightmost available position
                    row[j], row[empty_pos] = '.', '#'
                    empty_pos -= 1
        
        # 2. Rotate the matrix 90 degrees clockwise
        # New dimensions: n x m
        # (r, c) in original becomes (c, (m-1) - r) in rotated
        rotated_box = [['' for _ in range(m)] for _ in range(n)]
        
        for r in range(m):
            for c in range(n):
                rotated_box[c][m - 1 - r] = boxGrid[r][c]
                
        return rotated_box

File: 1861-rotating-the-box/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<char>> rotateTheBox(vector<vector<char>>& boxGrid) {
        int m = boxGrid.size();
        int n = boxGrid[0].size();

        // 1. Apply gravity to each row independently
        for (int i = 0; i < m; ++i) {
            // emptyPos tracks the rightmost available spot for a stone to fall into
            int emptyPos = n - 1;
            for (int j = n - 1; j >= 0; --j) {
                if (boxGrid[i][j] == '*') {
                    // Obstacle: stones cannot pass. Reset emptyPos to the left of the obstacle.
                    emptyPos = j - 1;
                } else if (boxGrid[i][j] == '#') {
                    // Stone: move it to the emptyPos and update emptyPos
                    boxGrid[i][j] = '.';
                    boxGrid[i][emptyPos] = '#';
                    emptyPos--;
                }
            }
        }

        // 2. Rotate the box 90 degrees clockwise
        // Original (m x n) -> Result (n x m)
        vector<vector<char>> result(n, vector<char>(m));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                // Mapping: (r, c) -> (c, m - 1 - r)
                result[j][m - 1 - i] = boxGrid[i][j];
            }
        }

        return result;
    }
};